### PR TITLE
[AUDIO] Implement volume control support

### PR DESCRIPTION
**Pull request by oleg-dubinskiy**
## Purpose

Implement volume level changing for Aux/MidiOut/WaveOut devices. The volume control is represented the following WINMM functions:
- `auxGetVolume`,
- `auxSetVolume`,
- `midiOutGetVolume`,
- `midiOutSetVolume`,
- `waveOutGetVolume`,
- `waveOutSetVolume`,

which are calling the following messages appropriately:
- `AUXDM_GETVOLUME`,
- `AUXDM_SETVOLUME`,
- `MODM_GETVOLUME`,
- `MODM_SETVOLUME`,
- `WODM_GETVOLUME`,
- `WODM_SETVOLUME`.  

This fixes volume control for several 3rd-party programs (like Fox Audio Player 0.10.2 from Rapps, Winamp 2.95 with WaveOut plugin). However it does not fix changing the volume in system volume mixers (SndVol32, MMSys), since they are using their own functionality instead. They technically do the same things, but apart from the functions mentioned above.

JIRA issue: [CORE-14780](https://jira.reactos.org/browse/CORE-14780)

## TODO

- [ ] **Implement volume control for Legacy mode too. Currently, it is implemented only for MMixer mode.**
- [x] For now, volume is changing only in range from 51% to 100%, but in range from 0% to 50% it is always at minimal 1% level. **Fix it to change in the whole range from 0% to 100%, as it's done in Windows.** **Done now in my new #6950 PR**. :slightly_smiling_face: 

## Result

See the following videos, which demonstrates how it works before after my changes. At least Fox Audio Player 0.10.2 from Rapps and Winamp 2.95 with WaveOut plugin are now changing the volume correctly. :smiley: 

Before:

[Part 1.webm](https://github.com/reactos/reactos/assets/26385117/53e002ce-4a1c-492a-bfd2-6e8f8d79173c)

After:

[Part 2.webm](https://github.com/reactos/reactos/assets/26385117/0add60ea-249c-4cd5-bf7a-dfa89ec2b225)

I also made a clip with these two videos together, comparing before and after behaviour. Will upload it on YouTube. Unfortunately it's too big to attach here. :slightly_frowning_face:

### CODE REVIEW

**Comment by julenuri** [comment]
Question, isn't it too fast in the decreasing the volume? Shouldn't it be more progressive? :+1: Anyway, nice job.

**Comment by oleg-dubinskiy** [comment]
> Question, isn't it too fast in the decreasing the volume? Shouldn't it be more progressive? 👍 Anyway, nice job.

@julenuri Yeah, perhaps. Maybe each volume changing step is too big and should be x2 less, so we need to improve its calculation... But I'm not sure yet. Any opinions? :thinking:

**Comment by binarymaster** [comment]
> But I'm not sure yet. Any opinions?

What about API tests?

File: sdk/lib/drivers/sound/mmebuddy/mmewrap.c [hunk 1]
```diff
@@ -365,3 +365,85 @@ MmeGetPosition(
...
+
+    Result = GetSoundDeviceFunctionTable(SoundDevice, &FunctionTable);
+    if (!MMSUCCESS(Result))
+        return TranslateInternalMmResult(Result);
+
+    if (!FunctionTable->GetVolume)
+        return MMSYSERR_NOTSUPPORTED;
+
+    /* Call the driver */
+    Result = FunctionTable->GetVolume(SoundDeviceInstance, DeviceId, pdwVolume);
```
Comment (binarymaster) [inline sdk/lib/drivers/sound/mmebuddy/mmewrap.c:404]: error C4133: 'function': incompatible types - from 'PDWORD_PTR' to 'PDWORD'

**Comment by oleg-dubinskiy** [comment]
> > But I'm not sure yet. Any opinions?
> 
> What about API tests?

Just checked: we have an apitests for Get/SetVolume, but I did not run them yet.

File: sdk/lib/drivers/sound/mmebuddy/mmewrap.c (hunk 1 above)
Comment (oleg-dubinskiy) [reply sdk/lib/drivers/sound/mmebuddy/mmewrap.c]: Should be fixed now. :wink:

File: sdk/lib/drivers/sound/mmebuddy/auxiliary/auxMessage.c [hunk 2]
```diff
@@ -45,6 +45,22 @@ auxMessage(
                                                    Parameter2);
             break;
         }
+
+        case AUXDM_GETVOLUME :
```
Comment (JoachimHenze) [inline sdk/lib/drivers/sound/mmebuddy/auxiliary/auxMessage.c:49]:
```suggestion
        case AUXDM_GETVOLUME:
```

File: sdk/lib/drivers/sound/mmebuddy/auxiliary/auxMessage.c [hunk 3]
```diff
@@ -45,6 +45,22 @@ auxMessage(
...
+        {
+            Result = MmeGetVolume(AUX_DEVICE_TYPE,
+                                  DeviceId,
+                                  PrivateHandle,
+                                  &Parameter1);
+        }
+
+        case AUXDM_SETVOLUME :
```
Comment (JoachimHenze) [inline sdk/lib/drivers/sound/mmebuddy/auxiliary/auxMessage.c:57]:
```suggestion
        case AUXDM_SETVOLUME:
```

File: sdk/lib/drivers/sound/mmebuddy/midi/modMessage.c [hunk 4]
```diff
@@ -75,6 +75,21 @@ modMessage(
             break;
         }
 
+        case MODM_GETVOLUME :
```
Comment (JoachimHenze) [inline sdk/lib/drivers/sound/mmebuddy/midi/modMessage.c:78]:
```suggestion
        case MODM_GETVOLUME:
```

File: sdk/lib/drivers/sound/mmebuddy/midi/modMessage.c [hunk 5]
```diff
@@ -75,6 +75,21 @@ modMessage(
...
+        {
+            Result = MmeGetVolume(MIDI_OUT_DEVICE_TYPE,
+                                  DeviceId,
+                                  PrivateHandle,
+                                  &Parameter1);
+        }
+
+        case MODM_SETVOLUME :
```
Comment (JoachimHenze) [inline sdk/lib/drivers/sound/mmebuddy/midi/modMessage.c:86]:
```suggestion
        case MODM_SETVOLUME:
```

File: sdk/lib/drivers/sound/mmebuddy/mmewrap.c [hunk 6]
```diff
@@ -365,3 +365,85 @@ MmeGetPosition(
...
+{
+    MMRESULT Result;
+    PSOUND_DEVICE_INSTANCE SoundDeviceInstance;
+    PSOUND_DEVICE SoundDevice;
+    PMMFUNCTION_TABLE FunctionTable;
+
+    /* Sanity check */
+    SND_ASSERT(DeviceType == AUX_DEVICE_TYPE ||
+               DeviceType == MIDI_OUT_DEVICE_TYPE ||
+               DeviceType == WAVE_OUT_DEVICE_TYPE);
```
Comment (JoachimHenze) [inline sdk/lib/drivers/sound/mmebuddy/mmewrap.c:384]: Just a question out of curiosity: Was this ever hit during any of your real-world-tests?

File: sdk/lib/drivers/sound/mmebuddy/mmewrap.c [hunk 7]
```diff
@@ -365,3 +365,85 @@ MmeGetPosition(
     return Result;
 }
 
+
```
Comment (JoachimHenze) [inline sdk/lib/drivers/sound/mmebuddy/mmewrap.c:368]:
```suggestion
```
yes, I admit: I am nitpicking here. But just because I do know that sooner or later I will want to port your PR back ;-)

File: sdk/lib/drivers/sound/mmebuddy/mmewrap.c [hunk 8]
```diff
@@ -365,3 +365,85 @@ MmeGetPosition(
...
+    if (!FunctionTable->GetVolume)
+        return MMSYSERR_NOTSUPPORTED;
+
+    /* Call the driver */
+    Result = FunctionTable->GetVolume(SoundDeviceInstance, DeviceId, (PDWORD)pdwVolume);
+
+    return Result;
+}
+
+
```
Comment (JoachimHenze) [inline sdk/lib/drivers/sound/mmebuddy/mmewrap.c:409]:
```suggestion
```
yes, I admit: I am nitpicking here. But just because I do know that sooner or later I will want to port your PR back ;-)

File: sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c [hunk 9]
```diff
@@ -118,6 +118,22 @@ wodMessage(
             break;
         }
 
+        case WODM_GETVOLUME :
```
Comment (JoachimHenze) [inline sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c:121]:
```suggestion
        case WODM_GETVOLUME:
```

File: sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c [hunk 10]
```diff
@@ -118,6 +118,22 @@ wodMessage(
...
+        {
+            Result = MmeGetVolume(WAVE_OUT_DEVICE_TYPE,
+                                  DeviceId,
+                                  PrivateHandle,
+                                  &Parameter1);
+        }
+
+        case WODM_SETVOLUME :
```
Comment (JoachimHenze) [inline sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c:129]:
```suggestion
        case WODM_SETVOLUME:
```

Comment (JoachimHenze) [inline sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c:129]: Is it really intended to not have a break between `WODM_GETVOLUME` and `case WODM_SETVOLUME`?
If so, I would add a comment, something like `// fall-through intended`.
Otherwise ofc better to add the `break;` ;-)

File: sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c [hunk 11]
```diff
@@ -118,6 +118,22 @@ wodMessage(
...
+        {
+            Result = MmeSetVolume(WAVE_OUT_DEVICE_TYPE,
+                                  DeviceId,
+                                  PrivateHandle,
+                                  Parameter1);
+        }
+
         case DRV_QUERYDEVICEINTERFACESIZE :
```
Comment (JoachimHenze) [inline sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c:137]: Same here. Either add something like `// fall-through intended` or `break;`

File: sdk/lib/drivers/sound/mmebuddy/midi/modMessage.c [hunk 12]
```diff
@@ -75,6 +75,21 @@ modMessage(
...
         }
 
+        case MODM_GETVOLUME :
+        {
+            Result = MmeGetVolume(MIDI_OUT_DEVICE_TYPE,
+                                  DeviceId,
+                                  PrivateHandle,
+                                  &Parameter1);
+        }
+
```
Comment (JoachimHenze) [inline sdk/lib/drivers/sound/mmebuddy/midi/modMessage.c:85]: Same here. Either add something like `// fall-through intended`or `break;`

File: sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c (hunk 10 above)
Comment (oleg-dubinskiy) [reply sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c:129]: Ah, indeed! I forgot to add a break. Thank you! :sweat_smile:

File: sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c (hunk 11 above)
Comment (oleg-dubinskiy) [reply sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c:137]: Thanks, my mistake.

File: sdk/lib/drivers/sound/mmebuddy/mmewrap.c (hunk 6 above)
Comment (oleg-dubinskiy) [reply sdk/lib/drivers/sound/mmebuddy/mmewrap.c:384]: No, it was never hit for me at all, since it should never be hit, because it is not used for other device types. I kept this assert just for sure.

Comment (JoachimHenze) [reply sdk/lib/drivers/sound/mmebuddy/mmewrap.c:384]: Thank you for the answer! Resolved.

File: sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c (hunk 11 above)
Comment (oleg-dubinskiy) [reply sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c:139]: Fixed.

File: sdk/lib/drivers/sound/mmebuddy/midi/modMessage.c (hunk 12 above)
Comment (oleg-dubinskiy) [reply sdk/lib/drivers/sound/mmebuddy/midi/modMessage.c:86]: Fixed, thanks.

**Comment by Doug-Lyons** [comment]
Regression tests running now: https://build.reactos.org/#/builders/9/builds/36539

**Comment by HBelusca** [comment]
> (Maybe just use MIXERCONTROLDETAILS_SIGNED structure instead of MIXERCONTROLDETAILS_UNSIGNED? 🤔

Not sure why this would help.

File: dll/win32/wdmaud.drv/mmixer.c [hunk 13]
```diff
@@ -805,6 +805,123 @@ WdmAudGetWavePositionByMMixer(
...
+                               &MixLine,
+                               MIXER_OBJECTF_MIXER | MIXER_GETLINEINFOF_COMPONENTTYPE);
+    if (!MMSUCCESS(Result))
+        return TranslateInternalMmResult(Result);
+
+    MixLineControls.cbStruct = sizeof(MIXERLINECONTROLS);
+    MixLineControls.dwLineID = MixLine.dwLineID;
+    MixLineControls.dwControlType = MIXERCONTROL_CONTROLTYPE_VOLUME;
+    MixLineControls.cControls = 1;
+    MixLineControls.cbmxctrl = sizeof(MIXERCONTROL);
```
Comment (HBelusca) [inline dll/win32/wdmaud.drv/mmixer.c:894]:
```suggestion
    MixLineControls.cbmxctrl = sizeof(MixControl);
```

File: dll/win32/wdmaud.drv/mmixer.c [hunk 14]
```diff
@@ -805,6 +805,123 @@ WdmAudGetWavePositionByMMixer(
...
+                                   &MixLineControls,
+                                   MIXER_OBJECTF_MIXER | MIXER_GETLINECONTROLSF_ONEBYTYPE);
+    if (!MMSUCCESS(Result))
+        return TranslateInternalMmResult(Result);
+
+    /* Convert volume level to be set */
+    MixControlDetailsU[0].dwValue = LOWORD(dwVolume); // Left channel
+    MixControlDetailsU[1].dwValue = HIWORD(dwVolume); // Right channel
+
+    MixControlDetails.cbStruct = sizeof(MIXERCONTROLDETAILS);
```
Comment (HBelusca) [inline dll/win32/wdmaud.drv/mmixer.c:909]:
```suggestion
    MixControlDetails.cbStruct = sizeof(MixControlDetails);
```

File: dll/win32/wdmaud.drv/wdmaud.h [hunk 15]
```diff
@@ -144,6 +144,18 @@ WdmAudGetWavePositionByMMixer(
     IN  struct _SOUND_DEVICE_INSTANCE* SoundDeviceInstance,
     IN  MMTIME* Time);
 
+MMRESULT
+WdmAudGetVolumeByMMixer(
+    _In_ struct _SOUND_DEVICE_INSTANCE* SoundDeviceInstance,
```
Comment (HBelusca) [inline dll/win32/wdmaud.drv/wdmaud.h:149]: Please use `PSOUND_DEVICE_INSTANCE` instead of `struct _SOUND_DEVICE_INSTANCE*` in your new code
(except in the mmebuddy header, for the time being, since it appears the typedef that allows to use the shorter name is defined after the functions...)

File: sdk/lib/drivers/sound/mmebuddy/mmewrap.c [hunk 16]
```diff
@@ -365,3 +365,82 @@ MmeGetPosition(
...
+    if (!MMSUCCESS(Result))
+        return TranslateInternalMmResult(Result);
+
+    if (!FunctionTable->GetVolume)
+        return MMSYSERR_NOTSUPPORTED;
+
+    /* Call the driver */
+    Result = FunctionTable->GetVolume(SoundDeviceInstance, DeviceId, (PDWORD)pdwVolume);
+
+    return Result;
```
Comment (HBelusca) [inline sdk/lib/drivers/sound/mmebuddy/mmewrap.c:405]: Here there's no need to call the TranslateInternal helper?

**Comment by oleg-dubinskiy** [comment]
> > (Maybe just use MIXERCONTROLDETAILS_SIGNED structure instead of MIXERCONTROLDETAILS_UNSIGNED? 🤔
> 
> Not sure why this would help.

Indeed. Signed structure is used only for the controls which have negative values.
Just found out something: the problem is caused by wrong `VolumeData->InputSteppingDelta`, which is calculated as `0x10000 / Steps` (`Steps` is a total count of volume steps). After changing it to `0x5000 / Steps`, now the volume is changing in range from 0% to 50% and always is 100% in range from 51% to 100%. Then its calculation needs to be improved...

File: sdk/lib/drivers/sound/mmebuddy/mmewrap.c [hunk 17]
```diff
@@ -365,3 +365,82 @@ MmeGetPosition(
...
+    if (!MMSUCCESS(Result))
+        return TranslateInternalMmResult(Result);
+
+    if (!FunctionTable->SetVolume)
+        return MMSYSERR_NOTSUPPORTED;
+
+    /* Call the driver */
+    Result = FunctionTable->SetVolume(SoundDeviceInstance, DeviceId, (DWORD)dwVolume);
+
+    return Result;
```
Comment (HBelusca) [inline sdk/lib/drivers/sound/mmebuddy/mmewrap.c:445]: Same question

**Comment by oleg-dubinskiy** [comment]
> > > (Maybe just use MIXERCONTROLDETAILS_SIGNED structure instead of MIXERCONTROLDETAILS_UNSIGNED? 🤔
> > 
> > 
> > Not sure why this would help.
> 
> Indeed. Signed structure is used only for the controls which have negative values. Just found out something: the problem is caused by wrong `VolumeData->InputSteppingDelta`, which is calculated as `0x10000 / Steps` (`Steps` is a total count of volume steps). After changing it to `0x5000 / Steps`, now the volume is changing in range from 0% to 50% and always is 100% in range from 51% to 100%. Then its calculation needs to be improved...

Here's the code I'm talking about:
https://github.com/reactos/reactos/blob/bc2621812fdbdfaeb1d7d27b5b4d9963b6b90993/sdk/lib/drivers/sound/mmixer/controls.c#L210

File: dll/win32/wdmaud.drv/mmixer.c [hunk 18]
```diff
@@ -805,6 +805,123 @@ WdmAudGetWavePositionByMMixer(
...
+                                   MIXER_OBJECTF_MIXER | MIXER_GETLINECONTROLSF_ONEBYTYPE);
+    if (!MMSUCCESS(Result))
+        return TranslateInternalMmResult(Result);
+
+    MixControlDetails.cbStruct = sizeof(MIXERCONTROLDETAILS);
+    MixControlDetails.dwControlID = MixControl.dwControlID;
+    MixControlDetails.cChannels = MixLine.cChannels;
+    MixControlDetails.cMultipleItems = 0;
+    MixControlDetails.cbDetails = sizeof(MIXERCONTROLDETAILS_UNSIGNED);
+    MixControlDetails.paDetails = MixControlDetailsU;
```
Comment (HBelusca) [inline dll/win32/wdmaud.drv/mmixer.c:852]: According to MSDN
https://learn.microsoft.com/en-us/windows/win32/api/mmeapi/ns-mmeapi-mixercontroldetails
> For controls other than MIXERCONTROL_CONTROLF_MULTIPLE types, the size of this buffer is the product of the cChannels and cbDetails members of the MIXERCONTROLDETAILS structure. For controls other than MIXERCONTROL_CONTROLF_MULTIPLE types, the size of this buffer is the product of the cChannels and cbDetails members of the MIXERCONTROLDETAILS structure.

Here it seems that your array has a fixed size (two elements), but how does this relate to the `cChannels` value (which you set to be equal to `MixLine.cChannels`) above?

**Comment by HBelusca** [comment]
> > Indeed. Signed structure is used only for the controls which have negative values. Just found out something: the problem is caused by wrong `VolumeData->InputSteppingDelta`, which is calculated as `0x10000 / Steps` (`Steps` is a total count of volume steps). After changing it to `0x5000 / Steps`, now the volume is changing in range from 0% to 50% and always is 100% in range from 51% to 100%. Then its calculation needs to be improved...
> 
> Here's the code I'm talking about:
> 
> https://github.com/reactos/reactos/blob/bc2621812fdbdfaeb1d7d27b5b4d9963b6b90993/sdk/lib/drivers/sound/mmixer/controls.c#L210


Then my question is how this InputSteppingDelta, and Steps variables, relate to the 0-100% percentage and the associated sliding bar selection.

Let me also give you an example why this calculation (in that code you pointed) looks strange:
If Steps == 0 then InputSteppingDelta is "infinite".
If Steps == 1 then InputSteppingDelta is equal to that 0x10000 or 0x5000 constant (what does this constant correspond to?)
If Steps == 2 then InputSteppingDelta is equal to half of that constant.
Basically, as the code currently is, InputSteppingDelta scales as the inverse of Steps ....

**Review by JoachimHenze — APPROVED** [review]
Approved already. +1 to all suggestions that hermes did up to now. Many thanks to Doug for the already performed testman run.

**Comment by oleg-dubinskiy** [comment]
> > > Indeed. Signed structure is used only for the controls which have negative values. Just found out something: the problem is caused by wrong `VolumeData->InputSteppingDelta`, which is calculated as `0x10000 / Steps` (`Steps` is a total count of volume steps). After changing it to `0x5000 / Steps`, now the volume is changing in range from 0% to 50% and always is 100% in range from 51% to 100%. Then its calculation needs to be improved...
> > 
> > 
> > Here's the code I'm talking about:
> > https://github.com/reactos/reactos/blob/bc2621812fdbdfaeb1d7d27b5b4d9963b6b90993/sdk/lib/drivers/sound/mmixer/controls.c#L210
> 
> Then my question is how this InputSteppingDelta, and Steps variables, relate to the 0-100% percentage and the associated sliding bar selection.
> 
> Let me also give you an example why this calculation (in that code you pointed) looks strange: If Steps == 0 then InputSteppingDelta is "infinite". If Steps == 1 then InputSteppingDelta is equal to that 0x10000 or 0x5000 constant (what does this constant correspond to?) If Steps == 2 then InputSteppingDelta is equal to half of that constant. Basically, as the code currently is, InputSteppingDelta scales as the inverse of Steps ....

Uh, as it appears, `Steps` variable is not 0, not 1, and even not 2! It actually has the 16, 32 and 64 values, depending on the volume position.

File: sdk/lib/drivers/sound/mmebuddy/mmewrap.c (hunk 16 above)
Comment (oleg-dubinskiy) [reply sdk/lib/drivers/sound/mmebuddy/mmewrap.c:405]: No. This helper is generally used only in one failure case handling -  when any of parameters are invalid. It does not change anything in all other cases. All it does is only converting two return values (which are not returned by the calling function) to the general error code. For any other values, it simply returns the same result.
The function I'm calling in this case, has all parameters valid, so using the helper will not change anything.
See its implementation at https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/drivers/sound/mmebuddy/utility.c;hb=f69e256376e939e63029c8d8be8b97eaa206eec0#l132.

File: sdk/lib/drivers/sound/mmebuddy/mmewrap.c (hunk 17 above)
Comment (oleg-dubinskiy) [reply sdk/lib/drivers/sound/mmebuddy/mmewrap.c:445]: No.

File: dll/win32/wdmaud.drv/mmixer.c [hunk 19]
```diff
@@ -805,6 +805,123 @@ WdmAudGetWavePositionByMMixer(
...
+    _Out_ PDWORD pdwVolume)
+{
+    MMRESULT Result;
+    MIXERLINE MixLine;
+    MIXERCONTROL MixControl;
+    MIXERLINECONTROLS MixLineControls;
+    MIXERCONTROLDETAILS MixControlDetails;
+    MIXERCONTROLDETAILS_UNSIGNED MixControlDetailsU[2];
+
+    MixLine.cbStruct = sizeof(MIXERLINE);
```
[the remark this answers is not in the document]
Comment (oleg-dubinskiy) [reply dll/win32/wdmaud.drv/mmixer.c]: Done.

File: dll/win32/wdmaud.drv/mmixer.c [hunk 20]
```diff
@@ -805,6 +805,123 @@ WdmAudGetWavePositionByMMixer(
...
+                               &MixLine,
+                               MIXER_OBJECTF_MIXER | MIXER_GETLINEINFOF_COMPONENTTYPE);
+    if (!MMSUCCESS(Result))
+        return TranslateInternalMmResult(Result);
+
+    MixLineControls.cbStruct = sizeof(MIXERLINECONTROLS);
+    MixLineControls.dwLineID = MixLine.dwLineID;
+    MixLineControls.dwControlType = MIXERCONTROL_CONTROLTYPE_VOLUME;
+    MixLineControls.cControls = 1;
+    MixLineControls.cbmxctrl = sizeof(MIXERCONTROL);
```
[the remark this answers is not in the document]
Comment (oleg-dubinskiy) [reply dll/win32/wdmaud.drv/mmixer.c]: Done.

File: dll/win32/wdmaud.drv/wdmaud.h (hunk 15 above)
Comment (oleg-dubinskiy) [reply dll/win32/wdmaud.drv/wdmaud.h]: Done. Changed it in this file and in wdmaud.h header.

File: dll/win32/wdmaud.drv/mmixer.c (hunk 18 above)
Comment (oleg-dubinskiy) [reply dll/win32/wdmaud.drv/mmixer.c:852]: @HBelusca according to `MIXERLINE` structure documentation:
https://learn.microsoft.com/en-us/windows/win32/api/mmeapi/ns-mmeapi-mixerline
>Maximum number of separate channels that can be manipulated independently for the audio line. The minimum value for this field is 1 because a line must have at least one channel.
Most modern audio cards for personal computers are stereo devices; for them, the value of this member is 2.
Channel 1 is assumed to be the left channel; channel 2 is assumed to be the right channel.
A multichannel line might have one or more uniform controls (controls that affect all channels of a line uniformly) associated with it.

`cChannels` member of this structure contains the actual number of channels, and it is supposed to have the same purpose and same range of values (1, 2 or more) as `cChannels` member of `MIXERCONTROLDETAILS`. So do we actually need to always use a hardcoded value, even when the actual number of channels is different?

Comment (oleg-dubinskiy) [reply dll/win32/wdmaud.drv/mmixer.c:852]: The hardcoded value 1 is used only for `MIXERCONTROL_CONTROLTYPE_MUTE` or `MIXERCONTROL_CONTROLTYPE_MUX`, but not for `MIXERCONTROL_CONTROLTYPE_VOLUME`.
Just look at our MMSys code. Here's `cChannels` value is taken from `MIXERLINE`:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/cpl/mmsys/volume.c;hb=f69e256376e939e63029c8d8be8b97eaa206eec0#l134
And then it is stored in `MIXERCONTROLDETAILS.cChannels` member:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/cpl/mmsys/volume.c;hb=f69e256376e939e63029c8d8be8b97eaa206eec0#l184
same as in other functions.
It is only hardcoded here:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/cpl/mmsys/volume.c;hb=f69e256376e939e63029c8d8be8b97eaa206eec0#l83
but that is actually Mute Control, not Volume.
Similarly is in SndVol32, so why not do the same here?

Comment (oleg-dubinskiy) [reply dll/win32/wdmaud.drv/mmixer.c:852]: According to the docs you linked,
Снимок экрана от 2024-05-23 10-42-06
it can take a value from `MIXERLINE.cChannels`. This is only a valid for our case, because we are not using `MIXEROCNTROL_CONTROLTYPE_CUSTOM` or `MIXERCONTROL_CONTROLTYPE_UNIFORM`. We are using `MIXERCONTROL_CONTROLTYPE_VOLUME` here, as you can see in the code above.

Comment (oleg-dubinskiy) [reply dll/win32/wdmaud.drv/mmixer.c:852]: The only a thing I probably need to do, is multiply the value of `cbDetails` to `cChannels`, but then it will fail in MMixer, since that member is validated there.

Comment (HBelusca) [reply dll/win32/wdmaud.drv/mmixer.c:852]: > So do we actually need to always use a hardcoded value, even when the actual number of channels is different?

That's what I'm not fully sure indeed. The hardcoded two-element in the array looks strange to me, but I'm not fully versed in the multimedia/sound stuff to have a definite answer.

**Review by HBelusca — APPROVED** [review]
PR looks OK as it is now. The only thing not fully clear is the hardcoded two-element arrays (see discussion above).

**Review by TAN-Gaming — APPROVED** [review]
The code looks OK. `MixControlDetailsU[2]` should be described.

File: dll/win32/wdmaud.drv/mmixer.c (hunk 18 above)
Comment (oleg-dubinskiy) [reply dll/win32/wdmaud.drv/mmixer.c:852]: As you can see, I'm using two elements of the array for getting the values from/storing the values in two channels appropriately, left and right. Single structure has only one `dwValue` member, so I can't store both of them in one member or simply assign input `dwVolume` to/get output `pdwVolume` from it. I did not try to do it yet, but I'm not sure it is correct, since MSDN states the following:
> The low-order word of this location contains the left-channel volume setting, and the high-order word contains the right-channel setting.

See
https://learn.microsoft.com/en-us/windows/win32/api/mmeapi/nf-mmeapi-waveoutgetvolume
and
https://learn.microsoft.com/en-us/windows/win32/api/mmeapi/nf-mmeapi-waveoutsetvolume
for Get and Set functions appropriately.

**Comment by oleg-dubinskiy** [comment]
> PR looks OK as it is now. The only thing not fully clear is the hardcoded two-element arrays (see discussion above).

OK, so can I implement Legacy part later? (since it's not currently used and MMixer is enabled by default)

**Comment by oleg-dubinskiy** [comment]
Haha! Eureka! :smiley: 
I just found the source of problem of wrong volume scaling! It isn't caused by MMixer code, that works completely correctly! The problem is actually in our AC97 driver! It seems to write wrong output values for the starting volume position, due to which the scaling begins at 50% instead of 0%.
When using original AC97 driver from Windows XP/2003 (ac97intc.sys renamed to ac97.sys), the volume level chaning works as it should, starts at 0% and finishes at 100%. :slightly_smiling_face: 
Will try to fix our AC97 driver soon. :wink:

**Comment by oleg-dubinskiy** [comment]
> Haha! Eureka! 😃 I just found the source of problem of wrong volume scaling! It isn't caused by MMixer code, that works completely correctly! The problem is actually in our AC97 driver! It seems to write wrong output values for the starting volume position, due to which the scaling begins at 50% instead of 0%. When using original AC97 driver from Windows XP/2003 (ac97intc.sys renamed to ac97.sys), the volume level chaning works as it should, starts at 0% and finishes at 100%. 🙂 Will try to fix our AC97 driver soon. 😉

Our MMixer code actually retrieves those values from AC97 driver via KS using property request. So it definitely can't be anything other, I'm sure!

**Comment by oleg-dubinskiy** [comment]
> The code looks OK. `MixControlDetailsU[2]` should be described.

Done @TAN-Gaming. Added a small comments for both Get and Set functions. Are they OK or do you think they need to be improved? Thanks.

File: dll/win32/wdmaud.drv/mmixer.c (hunk 18 above)
Comment (HBelusca) [reply dll/win32/wdmaud.drv/mmixer.c:852]: In any case, I wonder what you see as the value of `MixLine.cChannels` when you run your code in ReactOS (or in Windows) with the debugger.

File: sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c [hunk 21]
```diff
@@ -118,6 +118,24 @@ wodMessage(
             break;
         }
 
+        case WODM_GETVOLUME:
+        {
+            Result = MmeGetVolume(WAVE_OUT_DEVICE_TYPE,
+                                  DeviceId,
+                                  PrivateHandle,
+                                  &Parameter1);
```
Comment (HBelusca) [inline sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c:126]: I'm quite certain there is a bug here and in the other invocation(s) of this MmeGetVolume in a similar context.
Indeed, as for example the MmeGetPosition etc, when the generic `wodMessage()` callback function is invoked by the caller, the `Parameter1` should already be set to the address of the data buffer that will receive the sought value (for example, the volume).

**Review by HBelusca — CHANGES_REQUESTED** [review]

**Comment by TAN-Gaming** [comment]
> Done @TAN-Gaming. Added a small comments for both Get and Set functions. Are they OK or do you think they need to be improved? Thanks.

The comments are LGTM.

File: dll/win32/wdmaud.drv/mmixer.c (hunk 18 above)
Comment (oleg-dubinskiy) [reply dll/win32/wdmaud.drv/mmixer.c:852]: OK, just debugged it. Its value always equals to `2`, as it's an actual number of channels (at least when calling `SetVolume` function):
`(dll/win32/wdmaud.drv/mmixer.c:903) MixLine.cChannels 2
(dll/win32/wdmaud.drv/mmixer.c:903) MixLine.cChannels 2
(dll/win32/wdmaud.drv/mmixer.c:903) MixLine.cChannels 2
(dll/win32/wdmaud.drv/mmixer.c:903) MixLine.cChannels 2
(dll/win32/wdmaud.drv/mmixer.c:903) MixLine.cChannels 2`
I'm sure it's correct value in this case.
Seems I should multiply `cbDetails` member on this `cChannels` if the value is greater than `1`, according to MSDN documentation, but then I need to update the check in the appropriate MMIXER routine too, otherwise it will fail.

File: sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c (hunk 21 above)
Comment (oleg-dubinskiy) [reply sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c:126]: Do you mean this `DWORD_PTR Parameter1` should be already a pointer? Judging by the documentation, it's true, so I should to simply cast `PDWORD` to `DWORD_PTR` I think. Then the problem will be solved.

Comment (oleg-dubinskiy) [reply sdk/lib/drivers/sound/mmebuddy/wave/wodMessage.c]: Done @HBelusca. See my updated changes. :slightly_smiling_face: 
Maybe I did not stuck on this problem yet, because `GetVolume` functions was not called in my case yet.

File: dll/win32/wdmaud.drv/mmixer.c (hunk 18 above)
Comment (oleg-dubinskiy) [reply dll/win32/wdmaud.drv/mmixer.c:852]: > Seems I should multiply cbDetails member on this cChannels if the value is greater than 1, according to MSDN documentation, but then I need to update the check in the appropriate MMIXER routine too, otherwise it will fail.

Seems in case I'll do it, I also should update all other occurrences of the similar code in another components (in a separate PR), to avoid failing that.

**Review by HBelusca — APPROVED** [review]
OK I won't block the PR further.
The hardcoded 2 (stereo) channel number needs to be investigated. May be worth trying some code snippets from there onto a real hardware Windows system with more than two audio channels (like dolby surround or whatever????)

File: dll/win32/wdmaud.drv/mmixer.c (hunk 18 above)
Comment (oleg-dubinskiy) [reply dll/win32/wdmaud.drv/mmixer.c:852]: > I wonder what would happen on a multi-stereo system on real hardware though

Unfortunately I currently have no hardware with compatible (non-HD audio) card. :slightly_frowning_face: MMixer mode isn't testable with HD audio, so it needs Legacy one to be implemented. I would test it if I'd have such a hardware, but why you are sure it can have a different result in ROS/XP/2003 on real hardware? :thinking:

Comment (HBelusca) [reply dll/win32/wdmaud.drv/mmixer.c:852]: I would suggest you ask on mattermost or discord. Maybe @DarkFire01 has something to test, who knows? :thinking: :joy:

Comment (oleg-dubinskiy) [reply dll/win32/wdmaud.drv/mmixer.c:852]: > I would suggest you ask on mattermost or discord. Maybe @DarkFire01 has something to test, who knows? 🤔 😂

OK, will do.

Comment (oleg-dubinskiy) [reply dll/win32/wdmaud.drv/mmixer.c:852]: > > I would suggest you ask on mattermost or discord. Maybe @DarkFire01 has something to test, who knows? 🤔 😂
> 
> OK, will do.

Done: https://chat.reactos.org/reactos/pl/1icf1z1rwpfu9pc7sk9ehbjqth. :wink:
